### ISSUE TITLE
ROW POLICY with "ALL EXCEPT" not working properly

### ISSUE THREAD

**Issue by realyota** [ClickHouse/ClickHouse#68601]
**Describe the unexpected behaviour**
When I add a row policy with the modifier 'TO ALL EXCEPT', the row policy that grants full row visibility to a specific role is not applied when data is selected by a user with that role."

**How to reproduce**
* Which ClickHouse server version to use
24.7.3 (Official build)

```
-- as admin
CREATE DATABASE test;

CREATE TABLE foo
(
    `a` Int32
)
ENGINE = MergeTree
ORDER BY tuple();

INSERT INTO foo VALUES (1),(2);

CREATE USER myuser IDENTIFIED WITH PLAINTEXT_PASSWORD BY '';

CREATE ROLE foo_a;

GRANT SELECT ON test.* TO foo_a

CREATE ROLE foo_b;

GRANT SELECT ON test.* TO foo_b;

CREATE ROW POLICY foo_limited ON test.foo FOR SELECT USING a <= 1 TO foo_a;

CREATE ROW POLICY foo_all ON test.foo FOR SELECT USING 1 TO ALL EXCEPT foo_a;

GRANT foo_a TO myuser;
GRANT foo_b TO myuser;

-- as myuser
$ clickhouse-client --user myuser

SELECT *
FROM test.foo

Query id: 2dc60634-217a-458a-91f2-5dec9a6b0f41

-- should be both rows since foo_b has foo_all row policy
   ┌─a─┐
1. │ 1 │
   └───┘

SHOW CURRENT ROLES

Query id: bd54debe-52e9-4c42-8aaf-b45071b498f5

   ┌─role_name─┬─with_admin_option─┬─is_default─┐
1. │ foo_a     │                 0 │          1 │
2. │ foo_b     │                 0 │          1 │
   └───────────┴───────────────────┴────────────┘

-- when set role is issued everything is fine

SET ROLE foo_b;

SELECT *
FROM test.foo

Query id: 0c693655-5856-4f53-bee6-7345ca8142c8

   ┌─a─┐
1. │ 1 │
2. │ 2 │
   └───┘
```

**Expected behavior**
The final scope of row policy should be combined set of all permissive row policies (combined using the boolean OR operator).
User should see all rows from that table since `foo_b` role has `foo_all` row policy applied.

**Comment by ringerc** [comment]
@groeneai This bug is pretty old - write a repro to verify that it still exists in current CH versions

**Comment by groeneai** [comment]
@ringerc Verified. It still reproduces on the current release, and the result is identical on the version in the report, so the behaviour has not changed since 2024.

Fixture from the issue body, unmodified. Official static builds, server config carrying CI's `enable_access_control_improvements.xml` values. `myuser` reading `test.foo`:

| session state | 26.8.1.2041 | 24.7.3.42 |
|---|---|---|
| default roles (`foo_a` and `foo_b` enabled) | `[1]` | `[1]` |
| after `SET ROLE foo_b` | `[1, 2]` | `[1, 2]` |
| after `SET ROLE foo_a` | `[1]` | `[1]` |

Two controls, in the same runs, same result on both versions:

* Replace `foo_all` with `CREATE ROW POLICY ... USING 1 TO foo_b` (direct role target, no `ALL EXCEPT`) and keep both roles enabled: `[1, 2]`. Permissive policies are OR-ed correctly, so the mixer is not the variable here.
* Keep `foo_a` granted but drop it from the enabled set (`ALTER USER myuser DEFAULT ROLE foo_b`): `[1, 2]`. The deciding input is the enabled role set, not the granted one.

Mechanism, byte-identical on both versions and on master (`src/Access/RolesOrUsersSet.cpp:231-247`): the `EXCEPT` list is a veto, applied after the inclusion test, against the enabled roles.

```cpp
bool in_except_list = std::any_of(
    enabled_roles.begin(), enabled_roles.end(),
    [this](const UUID & enabled_role) { return except_ids.contains(enabled_role); });
return !in_except_list;
```

`foo_a` is enabled by default, so `foo_all` never reaches the filter mixer and `foo_limited` (`a <= 1`) is the only permissive filter left. The same `match` also selects quotas (`src/Access/QuotaCache.cpp:396`) and settings profiles (`src/Access/SettingsProfilesCache.cpp:173`), so the semantics are not row-policy-specific, and they predate the 2020 `dbms/` to `src/` move.

The behaviour is pinned by an existing test: `tests/integration/test_row_policy/test.py:877-905` (`test_policy_on_distributed_table_via_role`) builds the same shape, `all_data ... USING 1 TO ALL EXCEPT 'role1'` alongside `role1_data ... TO 'role1'` with `role1` as user1's default role, and asserts `[0, 2, 4, 6, 8]`, that is, the `ALL EXCEPT` policy not applying.

So the permissive `OR` the report expects is working. What the report reads as a mis-combination is `ALL EXCEPT <role>` meaning "no user for whom that role is currently enabled". The other reading would make the clause non-exclusive in practice, since `ALL` already includes everyone: any second enabled role would restore the policy the `EXCEPT` was meant to withhold.

No code change from me, and no new test, since the shape is already covered. If the semantics stand, the gap I see is documentation: the `TO` clause section says `ALL EXCEPT` "allows excluding some users from the all users list" and never states the enabled-role rule. I can send that one-paragraph clarification if you want it.

**Comment by ringerc** [comment]
@groeneai So to succinctly summarise, the problem is that the `EXCEPT` clause acts as a `RESTRICTIVE` policy even when `PERMISSIVE` is explicitly specified?

If so, respond including a code-block to "spell" the equivalent policy that used `EXCEPT` from this ticket using `RESTRICTIVE` policy instead of `EXCEPT`.

**Comment by groeneai** [comment]
@ringerc Not quite. `EXCEPT` is not a combining mode, it decides whether the policy applies to you at all. Nothing gets AND-ed: the `USING 1` filter is withheld from `myuser`, and `foo_limited` (`a <= 1`) is left as the only permissive filter for that session. The outcome in this ticket looks like a restrictive policy, but the two are not interchangeable, and the measurement below separates them.

The ticket's policy pair, rewritten without `EXCEPT`:

```sql
-- as in the ticket
CREATE ROW POLICY foo_limited ON test.foo FOR SELECT USING a <= 1 TO foo_a;
CREATE ROW POLICY foo_all     ON test.foo FOR SELECT USING 1      TO ALL EXCEPT foo_a;

-- same observable behaviour, no EXCEPT
CREATE ROW POLICY foo_all     ON test.foo FOR SELECT USING 1 TO ALL;
CREATE ROW POLICY foo_limited ON test.foo AS RESTRICTIVE FOR SELECT USING a <= 1 TO foo_a;
```

`myuser` reading `test.foo` on 26.8.1.2041, `groupArray(a)`, both spellings measured in the same run:

| session state | ticket (`TO ALL EXCEPT foo_a`) | rewrite (`AS RESTRICTIVE`) |
|---|---|---|
| default roles (`foo_a`, `foo_b`) | `[1]` | `[1]` |
| `SET ROLE foo_b` | `[1, 2]` | `[1, 2]` |
| `SET ROLE foo_a` | `[1]` | `[1]` |

Identical in all four combinations of `throw_on_unmatched_row_policies` and `users_without_row_policies_can_read_rows`.

Two caveats on the rewrite.

The permissive `USING 1 TO ALL` line is not optional. With the restrictive policy alone the rewrite matches only under the stock defaults, and flipping either setting away from its default breaks it: with `users_without_row_policies_can_read_rows` off, every session state returns `[]`; with `throw_on_unmatched_row_policies` on, which is what CI runs, the `SET ROLE foo_b` session fails with `ACCESS_DENIED` and ``myuser: Table `test`.`foo` has row policies, but none of them are for the current user``.

The two spellings diverge as soon as another permissive policy grants the same rows. Adding `CREATE ROW POLICY foo_b_all ON test.foo FOR SELECT USING 1 TO foo_b` and keeping both roles enabled gives `[1, 2]` under `TO ALL EXCEPT foo_a` and `[1]` under `AS RESTRICTIVE`, again in all four setting combinations. Restrictive is the stronger statement, so if the intent is "`foo_a` must never see `a > 1`", write it `AS RESTRICTIVE`. `TO ALL EXCEPT foo_a` only removes one route to those rows, and any other permissive policy still grants them.

`FiltersMixer::getResult` (`src/Access/RowPolicyCache.cpp:42-63`, identical at `v26.8.1.2041-lts` and on master) is where the OR and the AND happen, over the filters that survived the `TO` and `EXCEPT` match.

**Comment by clickgapai** [comment]
<details>
<summary>Reproducer</summary>

```sql
-- as a user with access management (e.g. default)
CREATE DATABASE test;
CREATE TABLE test.foo (a Int32) ENGINE = MergeTree ORDER BY tuple();
INSERT INTO test.foo VALUES (1),(2);

CREATE USER myuser IDENTIFIED WITH PLAINTEXT_PASSWORD BY '';
CREATE ROLE foo_a;
GRANT SELECT ON test.* TO foo_a;
CREATE ROLE foo_b;
GRANT SELECT ON test.* TO foo_b;

CREATE ROW POLICY foo_limited ON test.foo FOR SELECT USING a <= 1 TO foo_a;
CREATE ROW POLICY foo_all     ON test.foo FOR SELECT USING 1      TO ALL EXCEPT foo_a;

GRANT foo_a TO myuser;
GRANT foo_b TO myuser;

-- clickhouse-client --user myuser   (foo_a and foo_b are both default roles)
SELECT * FROM test.foo;    -- 1        <- foo_all withheld because foo_a is enabled
SET ROLE foo_b;
SELECT * FROM test.foo;    -- 1, 2     <- foo_a no longer enabled, foo_all applies

-- control: bind the same permissive policy to the role directly and the OR works
CREATE ROW POLICY OR REPLACE foo_all ON test.foo FOR SELECT USING 1 TO foo_b;
-- as myuser with both default roles: 1, 2
```
[▶ Run on ClickHouse Fiddle](https://fiddle.clickhouse.com/e464d9a9-f748-45f9-b7a6-e01b76828653)

</details>

**Results:**
- **Culprit PR:** not identified (details in Analysis metadata)
- **Reproduces on**: `master, 26.8, 26.7, 26.6, 26.3`
- **Tested versions**: `master, 26.8, 26.7, 26.6, 26.3`
- **Also tested** (unsupported, context only): `26.2` reproduces


CC component owner: @vitlibar @pufit

<details>
<summary>Analysis metadata</summary>

`comp-rbac` · Severity `P2` · Finding `h_issue_68601`

- Verified against `master` build `37b345f8ceda` (aarch64)
<details>
<summary>Introducing change: predates `v22.3.20.29` (oldest testable release)</summary>

- `v22.3.20.29` is the oldest release with a downloadable binary for this reproducer, and it already reproduces — the introducing change predates it and cannot be identified from release builds.

</details>



- Related: #93968, PR #97064, PR #95014


</details>

**Comment by ringerc** [comment]
@groeneai Try to restate the problem in clear simple English, succinctly. It might help to phrase this as a proposed improvement to the documentation.

**Comment by groeneai** [comment]
@ringerc Plain English: `TO` decides which policies you have, `AS` decides how the ones you have are combined. `TO ALL EXCEPT foo_a` means the policy does not apply to anyone who has `foo_a` enabled, so for `myuser` the `USING 1` policy is not one of their policies at all. There is nothing to OR it with, and the only permissive policy they do have is `a <= 1`, so they see one row. Nothing was AND-ed, a permissive alternative was withheld.

`EXCEPT` tests the roles currently **enabled** in the session, not the roles granted to the user. That is why `SET ROLE foo_b` changes the answer.

So I would call this a documentation gap rather than a server defect. Two edits would have prevented the ticket.

`TO` clause, after the existing `ALL EXCEPT` sentence:

> `ALL EXCEPT` is evaluated against the roles currently enabled in the session, not the roles granted to the user. A user who has an excluded role enabled is not covered by the policy, so that policy's condition is not among the conditions combined for them, and `SET ROLE` can change which policies apply.

`AS` clause, adding the missing qualifier to the formula:

```text
row_is_visible = (one or more of the permissive policies that apply to the user have non-zero conditions) AND
                 (all of the restrictive policies that apply to the user have non-zero conditions)
```

One more sentence worth having in the docs, because it is the practical trap: `TO ALL EXCEPT foo_a` is not a way to restrict `foo_a`. It withholds one permissive policy, and any other permissive policy that does apply hands the rows back. Measured on 26.8.1.2041: adding `USING 1 TO foo_b` returns both rows under `TO ALL EXCEPT foo_a`, but only one row when the same limit is written `AS RESTRICTIVE`. For "`foo_a` must never see `a > 1`", `AS RESTRICTIVE` is the spelling that holds.

I can send those doc edits as a PR if you want them.

**Comment by ringerc** [comment]
@groeneai Before raising a docs PR - does the `EXCEPT` clause's use of the currently-active role differ from how other row-security rule clauses work? Is that where the apparent confusion comes from here?

**Comment by groeneai** [comment]
@ringerc No, it does not differ. One function decides applicability for both halves of the clause: `RolesOrUsersSet::match(user_id, enabled_roles)` in `src/Access/RolesOrUsersSet.cpp:231-247`. The inclusion side (`:235-236`) and the `EXCEPT` veto (`:244-245`) test the same `enabled_roles` set.

Measured on 26.8.1.2041: add a third permissive policy `USING 1 TO foo_b` to the ticket's two, then `SET ROLE foo_a`, and `myuser` sees one row, not two. A plain `TO foo_b` with no `EXCEPT` also stops applying once `foo_b` is granted but not enabled. With both roles enabled the same fixture returns both rows, so the query does see that policy when it applies. The same predicate selects quotas (`QuotaCache.cpp:396`) and settings profiles (`SettingsProfilesCache.cpp:173`), so keying off enabled roles is the general rule, not something `EXCEPT` introduces.

On where the confusion comes from: it contributes, but I think it is the smaller of two reasons. `ALL EXCEPT foo_a` reads as a claim about the user, "has `foo_a` granted", while the code tests the session, "has `foo_a` enabled". `TO foo_a` carries the identical rule, but its wording does not invite that reading, and the docs do not help: the `TO` clause never mentions roles or enablement, only that `ALL EXCEPT` "allows excluding some users from the all users list".

The larger reason is the `AS` clause formula:

```text
row_is_visible = (one or more of the permissive policies' conditions are non-zero) AND
                 (all of the restrictive policies's conditions are non-zero)
```

It ranges over "the permissive policies" with no applicability qualifier, so read literally it promises that every permissive policy on the table is OR-ed for every user, which is exactly what this ticket expected. The qualifier appears two paragraphs earlier, "more than one policy enabled on the same table for the same user", and is then dropped from the formula.

So both edits are still worth making, and the `AS` one is the edit that would have prevented this ticket. If you want the PR: that prose lives in `src/Parsers/Access/ParserCreateRowPolicyQuery.cpp:421` and `:443`, and the documentation page is generated from it, so the change goes in that C++ file rather than the Markdown.

**Comment by ringerc** [comment]
@groeneai Sounds good, propose the appropriate docs/comments changes. Keep it minimal. Use clear language. Avoid references to bug numbers, files/lines, commits, or other noise the documentation reader does not need to know.

**Comment by groeneai** [comment]
@ringerc Acknowledged. I will send a minimal docs PR with two wording changes: one sentence in the `TO` clause stating that these clauses are evaluated against the roles enabled in the current session, not merely the roles granted, and the missing "that apply to the user" qualifier in the `AS` clause visibility formula. Plain prose, no issue, file or commit references in the doc text. I will post the link here.

**Comment by groeneai** [comment]
@ringerc https://github.com/ClickHouse/ClickHouse/pull/117624

The two wording changes I promised, plus a third I should flag: the same formula also implied that a user to whom no permissive policy applies sees no rows, when with the default server configuration the permissive condition is dropped instead and the applicable restrictive policies decide alone. That line was already wrong in that case, so I fixed it while rewriting it rather than leave a known error in place. I will drop it if you would rather keep the change to two.

The doc text names no settings, files or issues.

**Comment by ringerc** [comment]
I think this is probably fixed by the docs clarification in https://github.com/ClickHouse/ClickHouse/pull/117624 .

**Comment by groeneai** [comment]
#117624 merged on 2026-09-08 (`e1a9844c3413`), so the clarification is on master rather than pending. It went in with the third wording change I flagged in my previous comment, so that offer to drop it is moot.

No code change was needed and none was made: the combination logic itself works, and what the ticket ran into is `TO ... ALL EXCEPT` being matched against the roles enabled in the session rather than the roles granted to the user, which the docs now state. From my side this is a documentation-only outcome and the issue can be closed by a maintainer.
